core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -416,78 +351,116 @@ private long getCurrentBlockId() {
-    // The size of the last block, 0 if it is equal to the normal block size
-    long lastBlockSize = mFileLength % mBlockSize;
-    // If we are not in the last block or if the last block is equal to the normal block size,
-    // return the normal block size. Otherwise return the block size of the last block.
-    if (mFileLength - pos > lastBlockSize) {
-      return mBlockSize;
-    } else {
-      return lastBlockSize;
-    }
-  }
-
-  /**
    * Logs IO exceptions thrown in response to the worker cache request. If the exception is not an
    * expected exception, a warning will be logged with the stack trace.
    */
-  private void logCacheStreamIOException(IOException e) {
+  private void handleCacheStreamIOException(IOException e) throws IOException {
     if (e.getCause() instanceof BlockAlreadyExistsException) {
       LOG.warn(BLOCK_ID_EXISTS_SO_NOT_CACHED, getCurrentBlockId());
     } else {
       LOG.warn(BLOCK_ID_NOT_CACHED, getCurrentBlockId(), e);
     }
+    closeOrCancelCacheStream();
   }
 
   /**
-   * Similar to {@link #checkAndAdvanceBlockInStream()}, but a specific position can be specified
-   * and the stream pointer will be at that offset after this method completes.
+   * Only updates {@link #mCurrentCacheStream}, {@link #mCurrentBlockInStream} and
+   * {@link #mStreamBlockId} to be in-sync the current block (i.e. {@link #getCurrentBlockId()}).
+   * This function can be called multiple times without side effect. It is recommended to be
+   * invoked before every read and seek.
    *
-   * @param newPos the new position to set the stream to
-   * @throws IOException if the stream at the specified position cannot be opened
+   * @throws IOException if the next cache stream or block stream cannot be created
    */
-  private void seekBlockInStream(long newPos) throws IOException {
-    long oldBlockId = getCurrentBlockId();
-    mPos = newPos;
-    closeCacheStream();
+  private void updateStreams() throws IOException {
     long currentBlockId = getCurrentBlockId();
-
-    if (oldBlockId != currentBlockId) {
+    if (shouldUpdateStreams(currentBlockId)) {
+      // The following two function handle negative currentBlockId (i.e. the end of file)
+      // correctly.
       updateBlockInStream(currentBlockId);
-      // Reading next block entirely.
-      if (mPos % mBlockSize == 0 && mShouldCacheCurrentBlock) {
-        try {
-          WorkerNetAddress address = mLocationPolicy.getWorkerForNextBlock(
-              mContext.getAlluxioBlockStore().getWorkerInfoList(), getBlockSizeAllocation(mPos));
-          mCurrentCacheStream = createCacheStream(currentBlockId, getBlockSize(mPos), address);
-        } catch (IOException e) {
-          logCacheStreamIOException(e);
-          mShouldCacheCurrentBlock = false;
-        } catch (AlluxioException e) {
-          LOG.warn(BLOCK_ID_NOT_CACHED, currentBlockId, e);
-          mShouldCacheCurrentBlock = false;
+      updateCacheStream(currentBlockId);
```

Comment:
what if `updateBlockInStream` failed? should we still try to `updateCacheStream`?